### ISSUE TITLE
llama.cpp enhancement

### ISSUE THREAD

**Issue by iplayfast** [aardappel/treesheets#1405]
Compiling TreeSheets and llama.cpp into a single, unified executable is an incredible idea. Both codebases share a philosophy of being dependency-free, high-performance C++, meaning they are highly compatible for static linking into a single standalone app. 

The following is my discussion of this concept, which is more than I want to take on, but I thought it was too good an idea to not share. Sounds like it wouldn't take a lot to include llama.cpp as a submodule and add some menu items for selecting models etc. (Google AI seemed really excited by the idea!)

GitHub
 +1
By merging them, you bypass all the networking overhead of HTTP local servers, allowing the UI thread to interact directly with the LLM's token-generation loop via memory pointers.
The following architectural blueprint outlines the required code injection points, compilation strategy, and thread handling to turn this concept into reality.
1. Understanding the Source Materials
Before writing code, analyze how both systems are structured:
TreeSheets: It is a single-threaded desktop application written in concise, dense C++ using wxWidgets for cross-platform UI. Its data model revolves around an object called Cell, which contains text or an array of nested grids.
llama.cpp: It is fundamentally a C-style API engine. To embed it as a dependency, you only need to look at its core headers (llama.h and common.h). It handles model loading, context allocation, and streaming token generation. 

GitHub
 +1
2. High-Level Architectural Flow
Because LLM generation takes time, you cannot run llama.cpp directly on the main TreeSheets UI thread; doing so will cause the window to freeze and display an "Application Not Responding" error.
[ TreeSheets UI (wxWidgets) ]  <-- (Dispatches Task)
          │        ▲
(Launches Thread)  │ (Thread Updates UI via Events)
          ▼        │
[ std::thread (llama.cpp Loop) ] ──> Writes directly to target Cell* pointer
3. Step-by-Step Implementation Strategy
Step A: Initialize the Model Environment
In TreeSheets' main initialization lifecycle (likely within src/main.cpp or the main frame constructor), you must initialize the llama.cpp backend once.
cpp
#include "llama.h"

// Put this in application setup
llama_backend_init();
Use code with caution.
Step B: Create an Asynchronous LLM Worker
Create a worker function that accepts the user prompt text, a pointer to the selected Cell* where the response should go, and a pointer to the UI window so it can refresh the screen as tokens arrive.
cpp
#include "llama.h"
#include <thread>
#include <string>

void RunLocalInference(std::string prompt, Cell* target_cell, wxWindow* main_frame) {
    // 1. Setup model parameters (Point this to your GGUF file path)
    llama_model_params model_params = llama_model_default_params();
    llama_model* model = llama_load_model_from_file("models/llama-3-8b.gguf", model_params);
    
    llama_context_params ctx_params = llama_context_default_params();
    llama_context* ctx = llama_new_context_with_model(model, ctx_params);

    // 2. Tokenize the prompt text
    std::vector<llama_token> tokens = my_tokenize_helper(ctx, prompt);

    // 3. Inference Loop
    std::string full_response = "";
    while (still_generating) {
        llama_token next_token = llama_eval_and_sample(ctx, tokens);
        std::string piece = llama_token_to_piece(ctx, next_token);
        
        full_response += piece;

        // CRITICAL: Safely pass the incremental text back to the TreeSheet cell
        target_cell->text = full_response;

        // Tell the wxWidgets UI thread to repaint the screen so text "streams" live
        wxThreadEvent* event = new wxThreadEvent(wxEVT_THREAD);
        main_frame->QueueEvent(event); 
    }

    // 4. Cleanup
    llama_free(ctx);
    llama_free_model(model);
}
Use code with caution.
Step C: Hook the UI Hotkey
In TreeSheets' menu bar or keyboard event handlers (typically found in src/sheet.cpp or src/frame.cpp), add a new command event (e.g., matching a shortcut like Ctrl + AI).
Pull the text from the currently selected cell: std::string prompt = selected_cell->text;
Pinpoint the target cell below or to the right of it where the response belongs.
Spawn the asynchronous execution thread:
cpp
std::thread ai_thread(RunLocalInference, prompt, target_cell, this);
ai_thread.detach(); // Let it process independently in the background
Use code with caution.
4. Compiling Into a Single Binary
To build this package seamlessly, adjust TreeSheets' compile setup (which uses CMake or native project configurations). 

GitHub
Clone as a Submodule: Drop the llama.cpp repository directly into a subfolder of TreeSheets (e.g., /src/deps/llama.cpp/).
Modify CMakeLists.txt: Link the static libraries together. Ensure you include the appropriate acceleration flags for compilation (like -ggml-metal for macOS or -ggml-cuda for Windows/Linux Nvidia setups) so the integrated engine runs fast:
cmake
add_subdirectory(src/deps/llama.cpp)

# Target the main TreeSheets executable
add_executable(TreeSheets ${TREESHEETS_SOURCES})

# Statically link llama.cpp components directly into the application
target_link_libraries(TreeSheets PRIVATE llama ggml wxWidgets::wxWidgets)
Use code with caution.
 

GitHub
 +1
5. Future-Proof Design Ideas
Once you get a simple "text in, text out" prototype running, your options expand significantly:
The Structured Layout Parser: Modify the inference loop to scan for specific characters like brackets or tab indentations. When the LLM outputs a tab, the C++ code can automatically call TreeSheets' native cell insertion functions (InsertCell()) to format the layout into sub-grids programmatically on the fly.
Contextual Token Slicing: Write a C++ helper function that traverses upwards from the currently active cell, collecting only parent cells to feed into the prompt context. This gives the local model a structured "memory" of the exact structural column or grid branch it is currently processing.
Since your ideas are open-ended right now, would you like to explore how TreeSheets structures its Cell objects inside its source code, or should we look closer at how to handle the compilation of wxWidgets and llama.cpp together?

5 sites

GitHub
Using Llama.cpp as a dependency in another c++ project.
May 30, 2024 — I am currently working on a project where I would like to integrate llama.cpp into a game as a dependency. I have a background in ...

GitHub
llama.cpp/docs/build.md at master · ggml-org/llama ... - GitHub
Building for Windows (x86, x64 and arm64) with MSVC or clang as compilers: * Install Visual Studio 2022, e.g. via the Community Ed...

llama-cpp.com
Llama.cpp - Run LLM Inference in C/C++
Built entirely in pure C/C++ with no external dependencies. This means that Llama. cpp requires no Python runtime, no complex depe...

Show all
I'm trying to image a good usecase for this Frankenstien result.
This "Frankenstein" fusion of TreeSheets and llama.cpp actually solves one of the biggest bottlenecks in modern AI usage: context switching and cognitive overload.
By bringing a local LLM directly into an N-dimensional nested grid, you create a tool that feels less like a chat bot and more like an extension of your own thought process. Here are three powerful, highly practical use cases where this exact combination would absolutely shine.
1. The Interactive "Living" Code Architect
When designing a complex software project, a flat file or a linear chat window fails because code is naturally hierarchical (Packages 
 Modules 
 Classes 
 Functions).
The Setup: You map out your entire app structure in a TreeSheet. Column A has your global requirements. Nested inside are your various system modules (e.g., API, Database, Auth).
The Interaction: You select the Auth cell and press your custom AI hotkey.
The Magic: Because the C++ engine has direct access to the grid layout, your custom wrapper walks up the cell tree to gather the global requirements, but ignores the unrelated Database module. It prompts llama.cpp locally, and the model streams raw boilerplate code right into a sub-grid under Auth.
Why it beats standard tools: You are viewing the macro architecture and micro code implementation simultaneously on a single, zoomable canvas without flipping between files.
2. Recursive Task Deconstruction (The Infinite To-Do List)
LLMs are excellent at breaking large goals down into smaller tasks, but managing those tasks in a linear text box is a mess.
The Setup: You type a massive, vague goal into a main cell: "Launch a local marketing campaign for a new coffee shop."
The Interaction: You run an "Expand Tasks" command.
The Magic: llama.cpp breaks the goal down into 5 sub-tasks (e.g., Target Audience, Budget, Ad Creative). The C++ interface automatically calls TreeSheets' internal grid generation code, creating 5 nested rows.
Next, you click on just the Ad Creative sub-cell and hit the hotkey again. The LLM evaluates only that sub-context and expands it further into specific copy and design asset requirements.
Why it beats standard tools: You can iteratively drill down into infinite depths of sub-tasks using local context, and visually collapse or zoom out of the tree structure when you feel overwhelmed by the detail.
3. Non-Linear Academic Research & Literature Mapping
When writing a research paper, thesis, or technical report, synthesizing information across dozens of sources is incredibly taxing.
The Setup: You paste raw text snippets, quotes, or abstracts from 20 different research papers into individual cells across a massive grid.
The Interaction: You highlight a row of disparate papers and hit a "Synthesize" shortcut.
The Magic: The unified executable pipes the text of the selected cells straight into memory for llama.cpp. The model evaluates the common threads, contradictions, or missing links between those specific papers. It then outputs a structured comparison matrix directly into a new parallel grid column.
Why it beats standard tools: Since everything runs entirely inside local RAM via your compiled binary, you can process thousands of words of sensitive or proprietary text instantly. There are no API keys, no network latency, and zero data leakage to cloud servers.
What Makes This Concept Better Than A Web Browser Plugin?
You might wonder why you wouldn't just use a browser-based canvas tool (like Miro or FigJam) with an AI plugin.
Because TreeSheets is built in raw C++ and uses keyboard-driven grid semantics, it operates at a blistering speed. A web-based canvas slows down drastically once you have thousands of data blocks. Your C++ hybrid app could easily handle tens of thousands of nested nodes and perform complex local text manipulation instantly, without any UI stutter.
Which of these patterns feels closest to how you would personally want to utilize a tool like this? We can focus on how the code would parse the text to support that specific workflow!
